retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt

```diff
@@ -0,0 +1,225 @@
+          newRequest.header("Authorization", "$tokenType $accessToken")
+        }
+      }
+
+      return chain.proceed(newRequest.build())
+    }
+
+    private fun responseCount(response: Response): Int
+    {
+      var responseHolder: Response? = response
+      var result = 1
+
+      while (result <= MAX_RETRIES && responseHolder?.priorResponse() != null)
+      {
+        result++
+        responseHolder = response.priorResponse()
+      }
+
+      return result
+    }
+  }
+
+  private val httpAuthClient: OkHttpClient by lazy {
+    OkHttpClient.Builder()
+        .connectTimeout(connectTimeout, TimeUnit.MILLISECONDS)
+        .readTimeout(readTimeout, TimeUnit.MILLISECONDS)
+        .writeTimeout(writeTimeout, TimeUnit.MILLISECONDS)
+        .addNetworkInterceptor(TokenAuthenticatorInterceptor())
+        .build()
+  }
+
+  private val authService: AuthAPI? by lazy {
+    authProvider.run {
+      val serviceBuilder = Retrofit
+          .Builder()
+          .baseUrl(getAuthRoute())
+          .client(httpAuthClient)
+
+      converterFactories.forEach { converterFactory ->
+        serviceBuilder.addConverterFactory(converterFactory)
+      }
+
+      return@run serviceBuilder.build().create(AuthAPI::class.java)
+    }
+  }
+
+
+  /**
+   * Method to login the user with the implemented authenticator.
+   *
+   * @param[username] the username of the user to connect.
+   * @param[password] the password of the user to connect.
+   *
+   * @return the success of the connection.
+   */
+  @WorkerThread
+  protected fun loginUser(username: String, password: String): ResponseWithError<AccessToken, ErrorResponse>?
+  {
+    val responseWithError: ResponseWithError<AccessToken, ErrorResponse>? = try
+    {
+      executeAuth(authService?.authToken(authProvider.getAuthRoute(), username, password))
+    }
+    catch (exception: IOException)
+    {
+      warn("Call of login failed with exception: ${exception.printStackTrace()}")
+      null
+    }
+    catch (exception: JacksonExceptions.JacksonParsingException)
+    {
+      warn("Call of login failed with exception: ${exception.printStackTrace()}")
+      null
+    }
+
+    authProvider.setAccessToken(responseWithError?.successResponse)
+
+    return responseWithError
+  }
+
+  final override fun setupAuthenticator(): Authenticator?
+  {
+    return TokenAuthenticatorInterceptor()
+  }
+
+  final override fun setupFirstAppInterceptors(): List<Interceptor>?
+  {
+    return listOf(TokenAuthenticatorInterceptor())
+  }
+
+  final override fun shouldDoSecondCall(response: Response?, exception: Exception?): Boolean
+  {
+    return if (response?.code() == 401 || response?.code() == 403)
+    {
+      false
+    }
+    else
+    {
+      super.shouldDoSecondCall(response, exception)
+    }
+  }
+
+  @WorkerThread
+  @Throws(IOException::class, JacksonExceptions.JacksonParsingException::class)
+  private fun executeAuth(call: Call<AccessToken>?): ResponseWithError<AccessToken, ErrorResponse>?
```

Comment:
Another suggestion : 
```suggestion
return call?.request()?.let { request ->
      debug("Starting execution of auth call ${request.method()} to ${request.url()}")

      val newRequest = request.newBuilder().build()
      val response: Response? = httpAuthClient.newCall(newRequest).execute()
      val success = response?.isSuccessful
      val responseBody = response?.body()?.string()
      
      return@let ResponseWithError<AccessToken, ErrorResponse>().apply {
        if (success == true)
        {
          successResponse = mapResponseToObject(responseBody, AccessToken::class.java)
        }
        else
        {
          errorResponse = mapResponseToObject(responseBody, ErrorResponse::class.java)
        }
      }
    }
```